mopidy_youtube/youtube.py

```diff
@@ -0,0 +1,413 @@
+# -*- coding: utf-8 -*-
+
+import threading
+import datetime
+import pafy
+import requests
+import re
+import traceback
+from mopidy_youtube import logger
+
+# Wrapper for loading data of youtube videos/playlists via either the Youtube
+# API or pafy. It offers the possibility to load info in the background (using
+# threads), and use it later.
+#
+# eg
+#   video = youtube.Video.get("7uj0hOIm2kY")
+#   video.load_info()   # non-blocking
+#   ... later ...
+#   print video.length  # blocks until info arrives, if it hasn't already
+#
+
+yt_api_endpoint = 'https://www.googleapis.com/youtube/v3/'
+yt_key = 'AIzaSyAl1Xq9DwdE_KD4AtPaE4EJl3WZe2zCqg4'
+session = requests.Session()
+
+
+class API:
+    search_results = 15
+
+    # search for both videos and playlists using a single API call. fetches
+    # only title, thumbnails, channel (extra queries are needed for length and
+    # video_count)
+    #
+    @classmethod
+    def search(self, q):
+        query = {
+            'part': 'id,snippet',
+            'fields': 'items(id,snippet(title,thumbnails,channelTitle))',
+            'maxResults': self.search_results,
+            'type': 'video,playlist',
+            'q': q,
+            'key': yt_key
+        }
+        result = session.get(yt_api_endpoint+'search', params=query)
+        data = result.json()
+
+        def f(item):
+            if item['id']['kind'] == 'youtube#video':
+                obj = Video.get(item['id']['videoId'])
+            else:
+                obj = Playlist.get(item['id']['playlistId'])
+
+            obj._load_api_data(item)
+            return obj
+
+        return map(f, data['items'])
+
+
+# video or playlist
+class Entry(object):
```

Comment:
I would be more inclined if you would use https://github.com/tkem/cachetools. If you do stick to this one, please add separate tests...